### DISCUSSION TITLE
Version 2.3.6

### DISCUSSION THREAD

**Discussion by saran-t** [google-deepmind/mujoco#942] [Announcements]
> **NOTE:**  MuJoCo 2.3.6 is the last version to officially support Python 3.7.

## Models
1. Added [3x3x3 cube](https://github.com/deepmind/mujoco/blob/main/model/cube/cube_3x3x3.xml) example model. See [README](https://github.com/deepmind/mujoco/blob/0945af7b8a8cb504e308b29fdb3c55048350a365/model/cube/README.md) for details.

## Bug fixes
2. Fixed a bug that was causing an incorrect computation of the mesh bounding box and coordinate frame if the volume was invalid. In such case, now MuJoCo only accepts a non-watertight geometry if [`shellinertia`](https://mujoco.readthedocs.io/en/2.3.6/XMLreference.html#body-geom-shellinertia) is equal to `true`.
3. Fixed the sparse Jacobian multiplication logic that is used to compute derivatives for tendon damping and fluid force, which affects the behaviour of the [implicit and implicitfast integrators](https://mujoco.readthedocs.io/en/2.3.6/computation.html#geintegration).
4. Fixes to [`mj_ray`](https://mujoco.readthedocs.io/en/2.3.6/APIreference/APIfunctions.html#mj-ray), in line with geom visualisation conventions:
    - Planes and height-fields respect the `geom_group` and `flg_static` arguments. Before this change, rays would intersect planes and height-fields unconditionally.
    - `flg_static` now applies to all static geoms, not just those which are direct children of the world body.

## Plugins
5. Added touch-grid sensor plugin. See [documentation](https://github.com/deepmind/mujoco/blob/0945af7b8a8cb504e308b29fdb3c55048350a365/plugin/sensor/README.md) for details, and associated [touch_grid.xml](https://github.com/deepmind/mujoco/blob/0945af7b8a8cb504e308b29fdb3c55048350a365/model/plugin/touch_grid.xml) example model. The plugin includes [in-scene visualisation](https://youtu.be/0LOJ3WMnqeA).
6. Added Visualization tab to simulate UI, corresponding to elements of the [`visual`](https://mujoco.readthedocs.io/en/2.3.6/XMLreference.html#visual) MJCF element. After modifying values in the GUI, a saved XML will contain the new values. The modifiable members of [`mjStatistic`](https://mujoco.readthedocs.io/en/2.3.6/APIreference/APItypes.html#mjstatistic) ([`extent`](https://mujoco.readthedocs.io/en/2.3.6/XMLreference.html#statistic-extent), [`meansize`](https://mujoco.readthedocs.io/en/2.3.6/XMLreference.html#statistic-meansize) and [`center`](https://mujoco.readthedocs.io/en/2.3.6/XMLreference.html#statistic-center)) are computed by the compiler and therefore do not have defaults. In order for these attributes to appear in the saved XML, a value must be specified in the loaded XML.
7. Increased text width for UI elements in the default spacing. 

## General
8. Added [`mj_getState`](https://mujoco.readthedocs.io/en/2.3.6/APIreference/APIfunctions.html#mj-getstate) and [`mj_setState`](https://mujoco.readthedocs.io/en/2.3.6/APIreference/APIfunctions.html#mj-setstate) for getting and setting the simulation state as a concatenated vector of floating point numbers. See the [State](https://mujoco.readthedocs.io/en/2.3.6/computation.html#gestate) section for details.
9. Added [`mjContact.solreffriction`](https://mujoco.readthedocs.io/en/2.3.6/APIreference/APItypes.html#mjcontact), allowing different [`solref`](https://mujoco.readthedocs.io/en/2.3.6/modeling.html#csolver) parameters for the normal and frictional axes of contacts when using [elliptic friction cones](https://mujoco.readthedocs.io/en/2.3.6/XMLreference.html#option-cone). This attribute is required for elastic frictional collisions, see associated [example model](https://github.com/deepmind/mujoco/blob/0945af7b8a8cb504e308b29fdb3c55048350a365/test/engine/testdata/spin_recoil.xml) mimicking the spin-bounce recoil behaviour of [elastic rubber balls](https://www.youtube.com/watch?v=uFLJcRegIVQ&t=3s). This is an advanced option currently only supported by explicit [contact pairs](https://mujoco.readthedocs.io/en/2.3.6/XMLreference.html#contact-pair), using the [`solreffriction`](https://mujoco.readthedocs.io/en/2.3.6/XMLreference.html#contact-pair-solreffriction) attribute.
10. Added [`mjd_inverseFD`](https://mujoco.readthedocs.io/en/2.3.6/APIreference/APIfunctions.html#mjd-inversefd) for finite-differenced inverse-dynamics derivatives.
11. Added functions for operations on banded-then-dense “arrowhead” matrices. Such matrices are common when doing direct trajectory optimization. See [`mju_cholFactorBand`](https://mujoco.readthedocs.io/en/2.3.6/APIreference/APIfunctions.html#mju-cholfactorband) documentation for details.
12. Added [`mj_multiRay`](https://mujoco.readthedocs.io/en/2.3.6/APIreference/APIfunctions.html#mj-multiray) function for intersecting multiple rays emanating from a single point. This is significantly faster than calling [`mj_ray`](https://mujoco.readthedocs.io/en/2.3.6/APIreference/APIfunctions.html#mj-ray) multiple times.
13. Ray-mesh collisions are now up to 10x faster, using a bounding volume hierarchy of mesh faces.
14. Increased `mjMAXUIITEM` (maximum number of UI elements per section in Simulate) to 100.
15. Added [documentation](https://mujoco.readthedocs.io/en/2.3.6/programming/extension.html#exprovider) for resource providers.
16. Changed the formula for [`mju_sigmoid`](https://mujoco.readthedocs.io/en/2.3.6/APIreference/APIfunctions.html#mju-sigmoid), a finite-support sigmoid $s: \mathbf{R} \rightarrow [0, 1]$. Previously, the smooth part consisted of two stitched quadratics, once continuously differentiable. It is now a single quintic, twice continuously differentiable:

$$
s(x) =
     \begin{cases}
        0,                    &       & x \le 0  \\
        6x^5 - 15x^4 + 10x^3, & 0 \lt & x \lt 1  \\
        1,                    & 1 \le & x \qquad
     \end{cases}
$$

17. Added optional [`tausmooth`](https://mujoco.readthedocs.io/en/2.3.6/XMLreference.html#actuator-muscle-tausmooth) attribute to muscle actuators. When positive, the time-constant $\tau$ of muscle activation/deactivation uses [`mju_sigmoid`](https://mujoco.readthedocs.io/en/2.3.6/APIreference/APIfunctions.html#mju-sigmoid) to transition smoothly between the two extremal values given by the [Millard et al. (2013)](https://doi.org/10.1115/1.4023390) muscle model, within a range of width `tausmooth`. See [Muscle actuators](https://mujoco.readthedocs.io/en/2.3.6/modeling.html#cmuscle) for more details. Relatedly, [`mju_muscleDynamics`](https://mujoco.readthedocs.io/en/2.3.6/APIreference/APIfunctions.html#mju-muscledynamics) now takes 3 parameters instead of 2, adding the new smoothing-width parameter.
18. Moved public C macro definitions out of mujoco.h into a new public header file called [mjmacro.h](https://github.com/deepmind/mujoco/blob/0945af7b8a8cb504e308b29fdb3c55048350a365/include/mujoco/mjmacro.h). The new file is included by mujoco.h so this change does not break existing user code.
19. Added instrumentation for the [Address Sanitizer (ASAN)](https://clang.llvm.org/docs/AddressSanitizer.html) and [Memory Sanitizer (MSAN)](https://clang.llvm.org/docs/MemorySanitizer.html) to detect memory bugs when allocating from the mjData stack and arena.
20. Removed `pstack` and `parena` from the output of `mj_printData`, since these are implementation details of the `mjData` allocators that are affected by diagnostic paddings in instrumented builds.
21. Removed the `mj_activate` and `mj_deactivate` functions. These had been kept around for compatibility with old user code from when MuJoCo was closed source, but have been no-op functions since open sourcing.

<hr /><em>This discussion was created from the release <a href='https://github.com/deepmind/mujoco/releases/tag/2.3.6'>2.3.6</a>.</em>

**Comment by stm32f303ret6** [comment]
really nice feature the improvements of mj_multiRay and mj_ray, congrats
